src/java/org/apache/cassandra/index/sai/disk/vector/BruteForceRowIdIterator.java

```diff
@@ -61,59 +66,87 @@ public RowWithApproximateScore(int rowId, int ordinal, float appoximateScore)
     // more than the first handful of scores.
     // Priority queue with compressed vector scores
-    private final PriorityQueue<RowWithApproximateScore> approximateScoreQueue;
-    // Priority queue with full resolution scores
-    private final PriorityQueue<RowIdWithScore> exactScoreQueue;
+    private final SortingIterator<RowWithApproximateScore> approximateScoreQueue;
     private final JVectorLuceneOnDiskGraph.CloseableReranker reranker;
-    private final int topK;
-    private final int limit;
-    private int rerankedCount;
+    private boolean initialized;
+    private Object next;
 
     /**
      * @param approximateScoreQueue A priority queue of rows and their ordinal ordered by their approximate similarity scores
      * @param reranker A function that takes a graph ordinal and returns the exact similarity score
      * @param limit The query limit
      * @param topK The number of vectors to resolve and score before returning results
      */
-    public BruteForceRowIdIterator(PriorityQueue<RowWithApproximateScore> approximateScoreQueue,
+    public BruteForceRowIdIterator(SortingIterator<RowWithApproximateScore> approximateScoreQueue,
                                    JVectorLuceneOnDiskGraph.CloseableReranker reranker,
                                    int limit,
                                    int topK)
     {
+        super(RowIdWithScore::compare, new Object[topK]);
         this.approximateScoreQueue = approximateScoreQueue;
-        this.exactScoreQueue = new PriorityQueue<>(topK);
         this.reranker = reranker;
+        this.initialized = false;
+        this.next = null;
         assert topK >= limit : "topK must be greater than or equal to limit. Found: " + topK + " < " + limit;
-        this.limit = limit;
-        this.topK = topK;
-        this.rerankedCount = topK; // placeholder to kick off computeNext
     }
 
     @Override
-    protected RowIdWithScore computeNext() {
-        int consumed = rerankedCount - exactScoreQueue.size();
-        if (consumed >= limit) {
-            // Refill the exactScoreQueue until it reaches topK exact scores, or the approximate score queue is empty
-            while (!approximateScoreQueue.isEmpty() && exactScoreQueue.size() < topK) {
-                RowWithApproximateScore rowOrdinalScore = approximateScoreQueue.poll();
-                float score = reranker.similarityTo(rowOrdinalScore.ordinal);
-                exactScoreQueue.add(new RowIdWithScore(rowOrdinalScore.rowId, score));
+    public boolean hasNext()
+    {
+        if (next == null)
+        {
+            // Prepare/advance only when the next value is requested.
+            if (!initialized)
+            {
+                for (int i = 0; i < heap.length; ++i)
+                    heap[i] = getNextApproximateAndRerank();
+                heapify();
+                initialized = true;
+            }
+            else
+            {
+                replaceTop(getNextApproximateAndRerank());
```

Comment:
This changes the logic a bit. The original design drains the `limit` and then refills up to the `topk` count from the remaining approximate values, while the new design eagerly backfills rows. The assumed benefit of the original implementation is that we only need to read additional vectors from storage if there are shadowed primary keys.

I don't have performance data on this though. Do you feel strongly about making it eagerly back fill `RowIdWithScore` entries?

Reply:
Great, thanks